File: cpp/src/opcodes.cpp
/*
	Interpreter opcodes

	Copyright (c) 2022 Frank McIngvale, see LICENSE
*/
#include "opcodes.hpp"
#include "errors.hpp"
#include "interpreter.hpp"
#include <map>
using namespace std;

int64_t opcode_pack(uint8_t code, uint8_t A, uint16_t B, uint32_t C) {
	// C is max 20 bits
	if(C > 0x000fffff)
		throw LangError("C > 20 bits in opcode_pack()");

	return (int64_t)(code | (A<<8) | (B<<16) | (((int64_t)C)<<32));
}

void opcode_unpack(int64_t packed, uint8_t &code, uint8_t &A, uint16_t &B, uint32_t &C) {
	code = ((uint64_t)packed) & 0xff;
	A = (((uint64_t)packed) >> 8) & 0xff;
	B = (((uint64_t)packed) >> 16) & 0xffff;
	C = (((uint64_t)packed) >> 32) & 0x000fffff;
}

// use maps so this is order-independent
static map<string,uint8_t> NAME_TO_CODE { 
	{"FRAME-GET", OPCODE_FRAME_GET },
	{"FRAME-SET", OPCODE_FRAME_SET },
	{"JUMP-FORW", OPCODE_JUMP_FORW},
	{"JUMP-BACK", OPCODE_JUMP_BACK},
};

static map<uint8_t,string> CODE_TO_NAME {
	{OPCODE_FRAME_GET, "FRAME-GET"},
	{OPCODE_FRAME_SET, "FRAME-SET"},
	{OPCODE_JUMP_FORW, "JUMP-FORW"},
	{OPCODE_JUMP_BACK, "JUMP-BACK"},
};

uint8_t opcode_name_to_code(const char *name) {
	if(NAME_TO_CODE.find(name) == NAME_TO_CODE.end())
		throw LangError("No such opcode name: " + string(name));

	return NAME_TO_CODE[name];
}

const string& opcode_code_to_name(uint8_t code) {
	if(CODE_TO_NAME.find(code) == CODE_TO_NAME.end())
		throw LangError("No such opcode number: " + to_string(code));

	return CODE_TO_NAME[code];
}

// --- opcode functions ---

// A = number of frames up (0 means my frame); B = index of var in outer frame
void opcode_FRAME_GET(Interpreter *intr, uint8_t levels, uint16_t index, uint32_t _unused) {
	if(!intr->cur_framedata)
		throw LangError("opcode FRAME-GET called on null frame");

	intr->push(intr->cur_framedata->getFrameObj(levels, index));
}

// A = number of frames up (0 means my frame); B = index of var in outer frame
void opcode_FRAME_SET(Interpreter *intr, uint8_t levels, uint16_t index, uint32_t _unused) {
	if(!intr->cur_framedata)
		throw LangError("opcode FRAME-SET called on null frame");

	auto obj = intr->pop(); // get obj to be stored
	intr->cur_framedata->setFrameObj(levels, index, obj);
}

void do_opcode_JUMP(Interpreter *intr, int32_t offset) {
	if(!intr->code)
		throw LangError("opcode JUMP called but no code loaded??");

	auto newpos = intr->codepos + offset;
	if(newpos < 0 || newpos >= intr->code->size())
		throw LangError("JUMP out of bounds");

	intr->codepos = newpos;
}

void opcode_JUMP_FORW(Interpreter *intr, uint8_t _A, uint16_t _B, uint32_t offset) {
	do_opcode_JUMP(intr, offset);
}

void opcode_JUMP_BACK(Interpreter *intr, uint8_t _A, uint16_t _B, uint32_t offset) {
	do_opcode_JUMP(intr, -((int32_t)offset));
}

// make sure order matches ordering of values!
std::vector<opcode_func> OPCODE_FUNCTIONS {
	opcode_FRAME_GET, opcode_FRAME_SET, opcode_JUMP_FORW, opcode_JUMP_BACK };

File: csharp/opcodes.cs

/*
	Interpreter opcodes

	Ported from C++ implementation. See documentation there.

	Copyright (c) 2022 Frank McIngvale, see LICENSE
*/
#nullable enable

using System;
using System.IO;
using System.Collections.Generic;

public class Opcodes {
	// make sure these match c++ reference
	const byte OPCODE_FRAME_GET = 0; 
	const byte OPCODE_FRAME_SET = 1; 
	const byte OPCODE_JUMP_FORW = 2;
	const byte OPCODE_JUMP_BACK = 3;

	// see c++ notes for why this is a signed long return
	public static long opcode_pack(byte code, byte A, ushort B, uint C) {
		if(C > 0x000fffff)
			throw new LangError("C > 20 bits in opcode_pack()");

		return (long)code | ((long)A<<8) | ((long)B<<16) | (((long)C)<<32);
	}

	public static void opcode_unpack(long packed, out byte code, out byte A, out ushort B, out uint C) {
		code = (byte)(((ulong)packed) & 0xff);
		A = (byte)((((ulong)packed) >> 8) & 0xff);
		B = (ushort)((((ulong)packed) >> 16) & 0xffff);
		C = (uint)((((ulong)packed) >> 32) & 0x000fffff);
	}

	public static Dictionary<string,byte> NAME_TO_CODE = 
		new Dictionary<string,byte> { 
			{"FRAME-GET", OPCODE_FRAME_GET},
			{"FRAME-SET", OPCODE_FRAME_SET},
			{"JUMP-FORW", OPCODE_JUMP_FORW},
			{"JUMP-BACK", OPCODE_JUMP_BACK},
		};

	public static Dictionary<byte,string> CODE_TO_NAME = 
		new Dictionary<byte,string> { 
			{OPCODE_FRAME_GET, "FRAME-GET"},
			{OPCODE_FRAME_SET, "FRAME-SET"},
			{OPCODE_JUMP_FORW, "JUMP-FORW"},
			{OPCODE_JUMP_BACK, "JUMP-BACK"},
		};

	public static byte opcode_name_to_code(string name) {
		if(NAME_TO_CODE.ContainsKey(name))
			return NAME_TO_CODE[name];
		else
			throw new LangError("No such opcode name: " + name);
	}

	public static string opcode_code_to_name(byte code) {
		if(CODE_TO_NAME.ContainsKey(code))
			return CODE_TO_NAME[code];
		else
			throw new LangError("No such opcode number: " + code.ToString());
	}

	// opcode functions -- put into an array
	public static void op_FRAME_GET(Interpreter intr, byte levels, ushort index, uint _unused) {
		if(intr.framedata == null)
			throw new LangError("FRAME-GET called with null frame");
		
		intr.push(intr.framedata.getFrameObj(levels, index));
	}

	public static void op_FRAME_SET(Interpreter intr, byte levels, ushort index, uint _unused) {
		if(intr.framedata == null)
			throw new LangError("FRAME-GET called with null frame");
		
		var obj = intr.pop();
		intr.framedata.setFrameObj(levels, index, obj);
	}
	
	public static void do_op_JUMP(Interpreter intr, int offset) {
		if(intr.code == null)
			throw new LangError("JUMP called without code?");

		var pos = intr.codepos + offset;
		if(pos < 0 || pos >= intr.code.Count)
			throw new LangError("JUMP out of bounds");

		intr.codepos = pos;
	}

	public static void runOpcode(Interpreter intr, LangOpcode opcode) {
	
		switch(opcode.code) {
			case OPCODE_FRAME_GET:
				op_FRAME_GET(intr, opcode.A, opcode.B, opcode.C);
				break;

			case OPCODE_FRAME_SET:
				op_FRAME_SET(intr, opcode.A, opcode.B, opcode.C);
				break;
				
			case OPCODE_JUMP_FORW:
				do_op_JUMP(intr, (int)opcode.C);
				break;

			case OPCODE_JUMP_BACK:
				do_op_JUMP(intr, -((int)opcode.C));
				break;
				
			default:
				throw new LangError("Bad opcode: " + opcode.code.ToString());
		}
	}
}
